dosbox-staging/dosbox-staging
# Further mixer fixes and improvements mega PR #2

# Description

- Fix mixer time stretching regression (see first commit for details). It's relatively hard to trigger, but a regression nonetheless.

- The LERP upsampler can now also downsample if required. This resampling mode was introduced specifically for the default `sb_filter = modern` mode, which emulates the old DOSBox linear interpolation "resampler."

- Previously, the upsampler did nothing if the channel rate exceeded the host rate. This resulted in lower-pitched audio, e.g., when a program uses the maximum native 45,454 Hz Sound Blaster rate and the host rate is 44,100 Hz. Or if you set a 22,050 Hz host rate and the SB rate is above that. Now that's fixed; in these scenarios, now we use the Speex resampler to do proper downsampling.

- Because of an oversight, all audio channels defaulted to the LERP upsampler. This generally worked out fine as long as your host rate was 44.1 kHz or 48 kHz. For example, the max GUS rate is 44.1 kHz, so if your host rate is 44.1 kHz, you could never hit the lower-pitched audio bug... unless you set a lower than 44.1k host rate, then you could 🥶 Anyway, the LERP upsampler is now fixed, but we also default to the Speex resampler on all channels as the default resampling method (the individual audio devices can override this, of course).

- Host sample rates can now be set continuously from 8,000 to 96,000 Hz. What's the bloody point of that, one might wonder? Well, many DOS devices have non-standard sample rates, e.g. the MT-32 is fixed 32,000 Hz, the OPL 49,716 Hz, the Sound Blaster up to 45,454 Hz, etc. By setting custom sample rates, you can render WAV files at an audio device's native rate without resampling. This might be useful for debugging (e.g., for _yours truly_). Sure, going up to 96 kHz is rather pointless, but FluidSynth supports it, so why not, plus we need an upper limit anyway. 96 kHz is as good as any.

- I've got rid of the `TwoPassSincResampler` in Game Blaster, PS/1 Audio, and Tandy 3-Voice, and I'm using our standard Speex resampler instead. This partially addresses https://github.com/dosbox-staging/dosbox-staging/issues/3595. See the commit messages for the details.

- The `nosound` option now works better because it renders the audio and progresses the "audio frames rendered" counters. Previously, programs that somehow timed themselves to these counters counters were completely stalled in "no sound" mode (e.g., FastTracker II). Additionally, now you can capture the audio to a WAV file even in "no sound" mode.

- `blocksize` can now be set to non-power-of-two values too. It seems SDL negotiates such buffer sizes on certain audio cards.

# Manual testing

Executed all the [audio emulation test cases](https://github.com/dosbox-staging/dosbox-staging/wiki/Audio-emulation-tests), and the following additional tests:

- Start DOSBox normally without specifying `rate` in the config and confirm in the logs that 48 kHz host sample rate is being used. Run `PT.BAT` and `CM.BAT` to play the same music in PT4DOS (SB Pro) and CapaMod (GUS), respectively, to familiarise yourself how the music should sound when played back at normal speed.

- Set `rate = 22050` and confirm the music plays at normal speed in both programs, not at half-speed, like `main`. 

- Set `nosound = on` in the config and confirm that:
  - Both programs show the playback animations, and are not stalled like `main`.
  - Capturing the audio records the music at the correct speed.

- Confirm arbitrary rates can be set both (e.g., `rate = 33333`) both with `negotiate` enabled or disabled. Confirm audio capturing works at these arbitraty rates (e.g., load the WAV into Audacity and check the sample rate).

- Confirm `stype = sb16` works with:
  - `sb_filter = auto` and 8000, 22050, 44100, 48000, and 96000 rates.
  - `sb_filter = modern` and the same rates.
  - `sb_filter = off` and the same rates.

- Repeat the tests with `sbtype = sbpro2`.

- Confirm 8000 Hz and 96000 Hz sample rates work, and that reverb and chorus work fine at these rates too (the audio will sound muffled on 8000 Hz; that's normal).

- Confirm the below audio devices work fine at 8 Khz and 96 Khz (these devices render directly at the mixer rate without resampling at the mixer level).
  - CD Audio, Covox, Disney, Game Blaster, GUS, Innovation, PC Speaker (both modes), PS1 Audio, ReelMagic, Tandy

- Confirm the mixer is not reinitialised when changing any of these settings at runtime:
  - `chorus`, `compressor`, `crossfeed`, `reverb`

- Change the above settings at runtime and confirm the sound output is as expected.


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:
